### DISCUSSION TITLE
Sorting of Component Browser Entries

### DISCUSSION THREAD

**Discussion by MichaelMauderer** [enso-org/enso#5649] [Work In Progress - GUI] [answered]
After some initial feedback, I looked at one of the currents of irregular sorting. Specifically, we had the case where `File list_immediate_children` was sorted higher than `File list`. 

It does seem that [our scoring algorithm](https://github.com/enso-org/enso/blob/a3cd542db7c8c8dafdd1323c32752dfc85cd8a36/lib/rust/fuzzly/src/metric.rs#L97) does NOT consider the placement of the match in the target text. That is, whether `list` is found at the end of the string, or at the start of the string, is not considered at all. 

The only place where we do consider the placement is the search pattern. So, a match against the pattern "l**is**t"  scores lower than "**li**s" or "li**st**", but the placement of the match in the target text is not relevant. 
This behaviour actually contradicts the[ docs on the metric](https://github.com/enso-org/enso/blob/a3cd542db7c8c8dafdd1323c32752dfc85cd8a36/lib/rust/fuzzly/src/metric.rs#L74), which states that the matching algorithm should consider the 
`how far is first matched char from text beginning and last character from text ending`. 

I think both of these should be considered, as both of them make sense, so I would propose we add both with separate configurable weights that can be tweaked.  

@farmaazon

**Comment by farmaazon** [comment, marked as answer]
> It does seem that [our scoring algorithm](https://github.com/enso-org/enso/blob/a3cd542db7c8c8dafdd1323c32752dfc85cd8a36/lib/rust/fuzzly/src/metric.rs#L97) does NOT consider the placement of the match in the target text.

Theoretically it should. The code is:

```rust
        let is_first_pattern_char = vertex.layer == 0;
        let is_last_pattern_char = text.len().checked_sub(1).contains(&vertex.layer);
        let first_char_bonus = if is_first_pattern_char {
            self.base_weight / (vertex.position_in_text as f32 + 1.0) * self.beginning_weight
        } else {
            0.0
        };
        let last_char_bonus = if is_last_pattern_char {
            self.base_weight / (text.len() - vertex.position_in_text) as f32 * self.ending_weight
        } else {
            0.0
        };
```

As you see, if we are measuring first char in the pattern, we can add "first_char_bonus" which is lessened by `vertex.position_in_text`, so the text which has pattern's beginning matching exactly with the text beginning should have this bonus bigger. Same for the last char in pattern.

(What looks suspicious, is the line `let is_last_pattern_char = text.len().checked_sub(1).contains(&vertex.layer);` I don't know why we take last char from _text_ instead of pattern).

**Comment by MichaelMauderer** [comment]
You are right. I was misled by the main checks, first looking at the pattern position of the match.

You are double right about the `text.len().checked_sub(1).contains(&vertex.layer);` it should be `pattern.len().checked_sub(1).contains(&vertex.layer);`. It then ranks the matches correctly.

**Reply by wdanilo** [reply]
@MichaelMauderer please remember about marking this issue as resolved (if it is answered and we are all on the same page) and connecting GH issue if needed :)
